pandas/core/reshape/pivot.py

```diff
@@ -345,12 +329,22 @@ def _all_key(key):
                 table_pieces.append(piece)
                 margin_keys.append(all_key)
         else:
-            margin = grand_margin
+            from pandas import DataFrame
+
             cat_axis = 0
             for key, piece in table.groupby(level=0, axis=cat_axis, observed=observed):
-                all_key = _all_key(key)
+                if len(cols) > 1:
+                    all_key = _all_key(key)
+                else:
+                    all_key = margins_name
                 table_pieces.append(piece)
-                table_pieces.append(Series(margin[key], index=[all_key]))
+                # GH31016 this is to calculate margin for each group, and assign
+                # corresponded key as index
+                transformed_piece = DataFrame(piece.apply(aggfunc)).T
+                transformed_piece.index = Index([all_key], name=piece.index.name)
```

Comment:
@charlesdong1991 what if piece.index is a MultiIndex?  .name here will be None.  Do we need to use .names somewhere?

Reply:
you are right, it will be `None` if it is a MultiIndex. Need to have a deeper look to see it is reasonable to use `.names` because if using it, then that's a frozen list iirc, and we could not assign to `index.name`.